### PR TITLE
TLS::Stream compatibility with the Boost.Asio extensible model

### PR DESCRIPTION
**Pull request by hrantzsch**
Improvements and fixes suggested in #2648.

- TLS::Stream async functions now also support completion tokens other than callback functions
- TLS::Stream::async_shutdown's internal handler wrapper now correctly forwards the executor
- add virtual d'tors to boost error_category subclasses

### CODE REVIEW

File: src/lib/tls/asio/asio_stream.h [hunk 1]
```diff
@@ -539,18 +548,23 @@ class Stream
        *                the handler must be:
```
Comment (hrantzsch) [inline src/lib/tls/asio/asio_stream.h:548]: Fix parameter name in comment

File: src/lib/tls/asio/asio_stream.h [hunk 2]
```diff
@@ -502,32 +508,35 @@ class Stream
...
+             {
+                using completion_handler_t = std::decay_t<decltype(completion_handler)>;
+
+                BOOST_ASIO_WRITE_HANDLER_CHECK(completion_handler_t, completion_handler) type_check;
+
+                boost::system::error_code ec;
+                tls_encrypt(bufs, ec);
+
+                if (ec)
+                {
```
Comment (hrantzsch) [inline src/lib/tls/asio/asio_stream.h:525]: run formatter

**Review by Tymolc — APPROVED** [review]
LGTM 👍

**Comment by hrantzsch** [comment]
Thank you again @finalglyph. I can't add you as a reviewer as it seems, but I'd be happy if you could give it another skim.

By the way, I mostly applied the changes you had already explained so nicely in the issue. If you like, feel free to directly open a PR in the future.

**Comment by hrantzsch** [comment]
@randombit none of the appveyor jobs on master runs with `--with-boost`, do they? From the [backport PR](https://github.com/randombit/botan/pull/2673) it seems Visual Studio is not happy about the changes, so I'll look into that again.

**Comment by randombit** [comment]
No the Boost builds were dropped when all old MSVC builds were, but that leaves a big hole in CI. I'll look at re-adding MSVC 2019 Boost build.

**Comment by randombit** [comment]
#2674 added back a Boost MSVC build

**Comment by finalglyph** [comment]
@hrantzsch, thank you for reading my lengthy explanations and preparing this pull request. The code changes look OK to me. (Actually, making the destructors of those `error_category` classes `virtual` seems unnecessary, but I suppose there are other reasons behind this.)

My server, which uses that `tls_stream_replacement` workaround I showed in #2648, hasn't had any issues (e.g. crashes) since I made the "executor forwarding" fix two weeks ago. This pull request basically "merges" the (latest) `tls_stream_replacement` code from #2648 into `Botan::TLS::Stream`, so I'm pretty confident it will fully solve #2648. Please close that issue once this pull request is merged.

Thanks again for these fixes! `Botan::TLS::Stream` now properly makes use of boost.asio's "extensible asynchronous model", so its interface is now as flexible as that of `boost::asio::ssl::stream`.

**Pull request merged**